pkg/scheduler/algorithm/priorities/resource_bin_packing.go

```diff
@@ -0,0 +1,240 @@
+}
+
+// ResourceBinPackingPriorityDefault creates a resourceBinPacking based prioity
+func ResourceBinPackingPriorityDefault() *ResourceBinPacking {
+	defaultResourceBinPackingPrioritizer := &ResourceBinPacking{
+		Resources: []Resources{},
+	}
+	return defaultResourceBinPackingPrioritizer
+}
+
+// ResourceBinPackingPriorityMap is a priority function that favors nodes that have higher utlization of scare resource.
+// It will detect whether the requested scarce resource is present on a node, and then calculate a score ranging from 0 to 10
+// based total utlization (best fit)
+// - If none of the scare resource are requested, this node will be given the lowest priority.
+// - If the scarce resource is requested, the larger the resource utlization ratio, the higher the node's priority.
+func (r *ResourceBinPacking) ResourceBinPackingPriorityMap(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (schedulerapi.HostPriority, error) {
+	var score int
+	node := nodeInfo.Node()
+	if len(r.Resources) == 0 {
+		return schedulerapi.HostPriority{}, fmt.Errorf("resource not defined")
+	}
+	if node == nil {
+		return schedulerapi.HostPriority{}, fmt.Errorf("node not found")
+	}
+	nodeScore := float64(0)
+	weightSum := 0
+
+	for _, elm := range r.Resources {
+		nodeScore += calculateScareResourceScore(nodeInfo, pod, elm.Resource) * float64(elm.Weight)
+		weightSum += elm.Weight
+	}
+	nodeScore = nodeScore / float64(weightSum)
+	if glog.V(10) {
+		glog.Infof(
+			"%v -> %v: ResourceBinPackingPriority, Score: (%d)", pod.Name, node.Name, score,
+		)
+	}
+
+	return schedulerapi.HostPriority{
+		Host:  node.Name,
+		Score: int(math.Round(nodeScore)),
+	}, nil
+
+}
+
+// calculateScareResourceScore returns total utlization of the scare resource on the node
+func calculateScareResourceScore(nodeInfo *schedulercache.NodeInfo, pod *v1.Pod, resource v1.ResourceName) float64 {
+	allocatable := nodeInfo.AllocatableResource()
+	used := nodeInfo.RequestedResource()
+	var requested schedulercache.Resource
+	var requestedInit schedulercache.Resource
+	switch resource {
+	case v1.ResourceCPU:
+		{
+			for i := range pod.Spec.Containers {
+				container := &pod.Spec.Containers[i]
+				cpu, _ := priorityutil.GetNonzeroRequests(&container.Resources.Requests)
+				requested.MilliCPU += cpu
+			}
+			for i := range pod.Spec.InitContainers {
+				container := &pod.Spec.InitContainers[i]
+				cpu, _ := priorityutil.GetNonzeroRequests(&container.Resources.Requests)
+				if requestedInit.MilliCPU < cpu {
+					requestedInit.MilliCPU = cpu
+				}
+			}
+			if requestedInit.MilliCPU > requested.MilliCPU {
+				requested.MilliCPU = requestedInit.MilliCPU
+			}
+
+			//println("CPU used: %d, requested: %d, available: %d",
+			//	nodeInfo.NonZeroRequest().MilliCPU, requested.MilliCPU, allocatable.MilliCPU)
+			return float64((nodeInfo.NonZeroRequest().MilliCPU+requested.MilliCPU)*schedulerapi.MaxPriority) / float64(allocatable.MilliCPU)
```

Comment:
what does `nodeInfo.NonZeroRequest().MilliCPU` this represent ?

Reply:
I believe this can only happen when using LimitRange

Reply:
Not really u can see the implementation [here] (https://github.com/kubernetes/kubernetes/blob/ed06cbe3e32370da07e9035f89a28521372c9392/pkg/scheduler/algorithm/priorities/util/non_zero.go#L40)

Reply:
Not really, here is the method implementation https://github.com/kubernetes/kubernetes/blob/ed06cbe3e32370da07e9035f89a28521372c9392/pkg/scheduler/algorithm/priorities/util/non_zero.go#L40